src/python/pants/engine/target.py

```diff
@@ -329,35 +329,61 @@ def _find_registered_field_subclass(
         return cast(Optional[Type[_F]], subclass)
 
     @final
-    def get(self, field: Type[_F]) -> _F:
+    def _maybe_get(self, field: Type[_F]) -> Optional[_F]:
+        result = self.field_values.get(field, None)
+        if result is not None:
+            return cast(_F, result)
+        field_subclass = self._find_registered_field_subclass(
+            field, registered_fields=self.field_types
+        )
+        if field_subclass is not None:
+            return cast(_F, self.field_values[field_subclass])
+        return None
+
+    @final
+    def __getitem__(self, field: Type[_F]) -> _F:
+        """Get the requested `Field` instance belonging to this target.
+
+        If the `Field` is not registered on this `Target` type, this method will raise a
+        `KeyError`. To avoid this, you should first call `tgt.has_field()` or `tgt.has_fields()`
+        to ensure that the field is registered, or, alternatively, use `Target.get()`.
+
+        See the docstring for `Target.get()` for how this method handles subclasses of the
+        requested Field and for tips on how to use the returned value.
+        """
+        result = self._maybe_get(field)
+        if result is not None:
+            return result
+        raise KeyError(
+            f"The target `{self}` does not have a field `{field.__name__}`. Before calling "
+            f"`my_tgt[{field.__name__}]`, call `my_tgt.has_field({field.__name__})` to "
+            f"filter out any irrelevant Targets or call `my_tgt.get({field.__name__})` to use the "
+            f"default Field value."
+        )
+
+    @final
+    def get(self, field: Type[_F], *, default_raw_value: Optional[Any] = None) -> _F:
```

Comment:
The Any is a shame, but I know of no way to deal with this without mypy supporting recursive self types. You can sort of hack this using a doubly parameterized Field type, with the concrete Field type and the value type as parameters, but mypy hits an internal bug when you do so.

Reply:
> I'm not a fan of forcing the keyword. Its makes the interface not like dict 

That's the precise reason I'm enforcing a kwarg: this interface is very similar to `dict` but is subtly different. Subtle differences like this are the origin of many bugs.

I believe it's going to be a common gotcha that people think the second parameter determines the default return value when in reality it sets the default value passed to the constructor for that Field. The kwarg reduces that risk for confusion.

> The Any is a shame, but I know of no way to deal with this without mypy supporting recursive self types. 

Agreed, but I can't think of any way around it as you describe. The important thing to me is that we return `_F`, which is what allows MyPy to know that `tgt.get(Compatibility).value` is `Tuple[str, ...]` whereas `tgt.get(ZipSafe).value` is `bool` because `Compatibility` and `ZipSafe` are fully typed classes.

This imprecise type hint is another reason I think the kwarg is worth the verbosity. We don't have MyPy/type hints to guide the user that this is wrong: 

```python
default_compatibility = Compatibility('CPython==3.6.*', address=tgt.address)
print(tgt.get(Compatibility, default_compatibility))
```

They would only know it's wrong at runtime because we would end up calling the constructor `Compatibility(raw_value=default_compatibility)`, which is wrong.